### PR TITLE
feat: anchored region revision

### PR DESCRIPTION
**Pull request by scomea**
# Description
This change is a major revision to the anchored region component which does the following:
- now only a single div
- more control over region scaling (can size based on content, anchor, or fill available space)
- tracks layout parent resizing (ie. the offet parent or document root)
- supports fixed placement (useful for breaking out of parent divs)
- we no longer track viewport scrolling, but developpers can update the offset themselves
- falls back to using getBoundingClientRect in scenarios where intersection observer does not work 
- enables authors to always/never use gbcr if the default behavior isn't ideal

## Motivation & context
- addressing feedback on the previous iteration

## Issue type checklist
- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

**Adding or modifying component(s) in `@microsoft/fast-components` checklist**
- [ ] I have added a new component
- [x] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## Process & policy checklist

- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

**Comment by scomea** [comment]
And yes, I'd love to come up with a better attribute name than "use-gbcr"...

File: packages/web-components/fast-components/src/anchored-region/resize-observer.ts [hunk 1]
```diff
@@ -1,17 +1,17 @@
-import { ResizeObserverEntry } from "./resize-observer-entry";
+// import { ResizeObserverEntry } from "./resize-observer-entry";
 
-export type ConstructibleResizeObserver = new (
-    callback: ResizeObserverCallback
-) => ResizeObserverClassDefinition;
+// export type ConstructibleResizeObserver = new (
```
Comment (nicholasrice) [inline packages/web-components/fast-components/src/anchored-region/resize-observer.ts:3]: Can this file be removed?

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts [hunk 2]
```diff
@@ -56,124 +65,250 @@ enum Location {
...
+    private useGbcrChanged(): void {
+        if ((this as any).$fastController.isConnected && this.initialLayoutComplete) {
+            this.initialize();
+        }
+    }
 
+    /**
+     * The HTML element being used as the anchor
+     *
+     * @ public
```
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts:274]:
```suggestion
     * @public
```

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts [hunk 3]
```diff
@@ -56,124 +65,250 @@ enum Location {
...
-            this.initialLayoutComplete = false;
-            this.reset();
-        }
+        this.reset();
     }
 
+    /**
+     * The HTML element being used as the viewport
+     *
+     * @ public
```
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts:285]:
```suggestion
     * @public
```

File: packages/web-components/fast-components/src/anchored-region/resize-observer.ts (hunk 1 above)
Comment (scomea) [reply packages/web-components/fast-components/src/anchored-region/resize-observer.ts]: yes, there was a redundant copy of those, gone.

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts [hunk 4]
```diff
@@ -376,41 +604,120 @@ export class AnchoredRegion extends FASTElement {
...
+            return null;
         }
-        this.requestLayoutUpdate();
+
+        this.regionDimension = {
+            height: regionRect.height,
+            width: regionRect.width,
+        };
+
+        // todo: scomea, try to avoid getting viewportRect when it is document root
```
Comment (nicholasrice) [inline packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts:686]: Is this still actionable?

Comment (scomea) [reply packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts:686]: Yes, I think there is potential perf optimization there, but deferring on it until later.

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts [hunk 5]
```diff
@@ -734,22 +1024,26 @@ export class AnchoredRegion extends FASTElement {
         desiredHorizontalPosition: AnchoredRegionHorizontalPositionLabel,
         nextPositionerDimension: Dimension
     ): void => {
+        let layoutParentWidth = 0;
+        if (this.offsetParent !== null) {
+            layoutParentWidth = this.offsetParent.clientWidth;
+        } else {
+            layoutParentWidth = document.body.clientWidth;
+        }
```
Comment (janechu) [inline packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts:1032]:
```suggestion
        const layoutParentWidth = this.offsetParent !== null
            ? this.offsetParent.clientWidth
            : document.body.clientWidth;
```

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts [hunk 6]
```diff
@@ -779,142 +1084,116 @@ export class AnchoredRegion extends FASTElement {
         desiredVerticalPosition: AnchoredRegionVerticalPositionLabel,
         nextPositionerDimension: Dimension
     ): void => {
+        let layoutParentHeight = 0;
+        if (this.offsetParent !== null) {
+            layoutParentHeight = this.offsetParent.clientHeight;
+        } else {
+            layoutParentHeight = document.body.clientHeight;
+        }
```
Comment (janechu) [inline packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts:1092]:
```suggestion
        const layoutParentHeight = this.offsetParent !== null
            ? this.offsetParent.clientHeight
            : document.body.clientHeight;
        let layoutParentHeight = 0;
```

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.spec.md [hunk 7]
```diff
@@ -191,6 +155,11 @@ A region that renders below the anchor until that space is less than 100px.
 </div>
 ```
 
+The anchored region can be configured to scale to either the size of the content in it's slot, to the size of the element it is anchored to, or the available space between the anchor and the edge of the viewport element by setting the vertical/horizontal scaling attribute:
+- "anchor" - size to match anchor
+- "fill" - size to match space between the anchor and the viewport edge
+- "content" - the default, matches the size of the content in the region's slot.
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/web-components/fast-foundation/src/anchored-region/anchored-region.spec.md:161]: possesive "region's" is correct

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts [hunk 8]
```diff
@@ -56,124 +65,250 @@ enum Location {
...
+     * @public
+     * @remarks
+     * HTML Attribute: use-gbcr
      */
-    @observable
-    public initialLayoutComplete: boolean = false;
+    @attr({ attribute: "use-gbcr" })
+    public useGbcr: GbcrUsage = "default";
+    private useGbcrChanged(): void {
+        if ((this as any).$fastController.isConnected && this.initialLayoutComplete) {
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/web-components/fast-foundation/src/anchored-region/anchored-region.ts:266]: Yes, that works.

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.spec.md [hunk 9]
```diff
@@ -233,34 +200,41 @@ NOTE: this component api will not be exposed outside of the fast-components pack
...
 ```
 
 ## Implementation
 
 ### States
 Layout update checks in the component happen when:
-- the viewport scrolls
 - intersection observer reports a collision with the viewport
-- resize observer reports a resize event on the anchor, the viewport or the component itself.
+- resize observer reports a resize event on the anchor, the viewport, the component's offsetParent or the component itself.
```
[the remark this answers is not in the document]
Comment (scomea) [reply packages/web-components/fast-foundation/src/anchored-region/anchored-region.spec.md]: added the ticks, but "component's" is correct for possessive

File: packages/web-components/fast-foundation/src/anchored-region/anchored-region.spec.md [hunk 10]
```diff
@@ -4,59 +4,23 @@
...
-
-A primary goal of the component was to enable authors to create responsive layouts without resorting to expensive [getBoundingClientRect()](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) calls. It instead depends on the more performant [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver) and [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) interfaces.
-
-For a more in-depth understanding of how this component works under the covers please refer to the [intersection observer api](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). 
-
 ### Use Cases
 
-It is envisioned that this component would be used as a building block for other components in this library (select, flyout, tooltip, etc) as well as being available for standalone use in responsive layouts.
+It is envisioned that this component would be used as a building block for other components in this library who need to position elements relative to another HTMLElement(select, flyout, tooltip, etc) as well as being available for standalone use in responsive layouts.
```
Comment (awentzel) [inline packages/web-components/fast-foundation/src/anchored-region/anchored-region.spec.md:9]: Would "that" be more grammatically correct than "who" need to position elements....

**Pull request merged**